Add alternative parameterization to negative binomial distribution #4126

ricardoV94 opened the pull request:
Changes allow the specification of the NegativeBinomial distribution in terms of p (the probability of observing a success in each trial), and n (the target number of successes). This parametrization gives the likelihood for y, the number of failures observed until a target number of successes is reached.

I changed the docstrings and distribution initialization in line with other distribution that have multiple parametrizations.

I didn't figure out yet how the tests work, so I did not implement it (hence the WIP)! This should be straightforward, however, since the scipy negativebinomial distribution is implemented in terms of n and p. Any guidance in how to proceed is much appreciated!

Pinging @twiecki 

Ref #4126

ricardoV94 committed:
Add alternative parameters p and n

ricardoV94 committed:
Update Release Notes

twiecki commented:
This is great -- thanks! Definitely tests are required, do you need help figuring that out?

ricardoV94 commented:
Yeah, I would need some help with that. I didn't find anything on the developer's guide about testing, and the test code is a bit opaque for me.

AlexAndorra commented:
Thanks for the PR @ricardoV94 ! For the tests, I think you can take inspiration from the tests of Negative Binomial done with the current parametrization. Doing them with the new one should already be a good sample of use-cases

ricardoV94 committed:
Add test

ricardoV94 commented:
@twiecki @AlexAndorra would the added test be enough?

MarcoGorelli commented:
IMO there should also be a test which hits the various error messages you've added, e.g. when you set both `alpha` and `n` - is that something you'd be happy to add? Feel free to ping if you want/need help

ricardoV94 commented:
@MarcoGorelli That makes sense.

I am not sure how to do it. Could you direct me to a snippet where this type of test is implemented? For example, the Beta distribution also raises this type of ValueErrors, but I couldn't find if it is being tested anywhere.

MarcoGorelli commented:
They do something similar in `pymc3/tests/test_starting.py`:

```python
    with raises(ValueError, match=r"Some variables not in the model: \['x2', 'y2'\]"):
        starting.allinmodel([x2, y2], model1)
    with raises(ValueError, match=r"Some variables not in the model: \['x2'\]"):
        starting.allinmodel([x2, y1], model1)
    with raises(ValueError, match=r"Some variables not in the model: \['x2'\]"):
        starting.allinmodel([x2], model1)
```

where `raises` was imported from `pytest` (see https://docs.pytest.org/en/stable/assert.html for more on this)

ricardoV94 committed:
Add test for invalid initializations

ricardoV94 commented:
@MarcoGorelli Thanks for the input. I added some tests that cover the new ValueError exceptions. What do you think?

MarcoGorelli commented:
@ricardoV94 nice work!

I think this can be simplified a bit using `pytest.mark.paramterize`, something like this (I've only written the first two test cases, but it should be easy to extend):
```python
    @pytest.mark.parametrize(
        "mu, n, alpha, expected",
        [
            (5, None, None, "Incompatible parametrization. Must specify either alpha or n."),
            (None, 2, 2, "Incompatible parametrization Can't specify both alpha and n."),
            # other two test cases go here
        ]
    )
    def test_negative_binomial_init_fail(self, mu, n, alpha, expected):
        with Model():
            with pytest.raises(ValueError, match=expected):
                x = NegativeBinomial("x", mu=mu, n=n, alpha=alpha)
```

MarcoGorelli commented on pymc3/distributions/discrete.py:
```diff
@@ -598,20 +598,60 @@ def NegBinom(a, m, x):
+                )
+        elif n is not None:
+            raise ValueError(
+                "Incompatible parametrization Can't specify both alpha and n."
```
missing full-stop?
```suggestion
                "Incompatible parametrization. Can't specify both alpha and n."
```

ricardoV94 committed:
Refactor tests with pytest.mark.parametrize

ricardoV94 commented:
@MarcoGorelli Thanks for the suggestion. I refactored the code with the `pytest.mark.parametrize` as you suggested. 

I also added a couple of tests for the different permutations of missing/over-specified parameters, do you think it is too much?

ricardoV94 committed:
Minor change

twiecki commented:
Thanks @ricardoV94! Is this your first PR?

ricardoV94 commented:
You're welcome. Yes it was the first one for me :)

twiecki commented:
Congratulations then! :)

On Thu, Oct 1, 2020 at 9:52 PM ricardoV94 <notifications@github.com> wrote:

> You're welcome. Yes it was the first one for me :)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/4134#issuecomment-702361670>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFETGC75NLNGNEERMDE4R3SITMYJANCNFSM4R2Z2X6A>
> .
>
